packages/angular_devkit/build_angular/src/builders/karma/find-tests.ts

```diff
@@ -6,55 +6,85 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import { existsSync } from 'fs';
-import * as glob from 'glob';
-import { basename, dirname, extname, join } from 'path';
-import { isDirectory } from '../../utils/is-directory';
+import { PathLike, constants, promises as fs } from 'fs';
+import glob, { hasMagic } from 'glob';
+import { basename, dirname, extname, join, relative } from 'path';
+import { promisify } from 'util';
+
+const globPromise = promisify(glob);
 
 // go through all patterns and find unique list of files
-export function findTests(patterns: string[], cwd: string, workspaceRoot: string): string[] {
-  return patterns.reduce((files, pattern) => {
-    const relativePathToMain = cwd.replace(workspaceRoot, '').substr(1); // remove leading slash
-    const tests = findMatchingTests(pattern, cwd, relativePathToMain);
-    tests.forEach((file) => {
-      if (!files.includes(file)) {
-        files.push(file);
-      }
-    });
+export async function findTests(
+  patterns: string[],
+  workspaceRoot: string,
+  projectSourceRoot: string,
+): Promise<string[]> {
+  const matchingTestsPromises = patterns.map((pattern) =>
+    findMatchingTests(pattern, workspaceRoot, projectSourceRoot),
+  );
+  const files = await Promise.all(matchingTestsPromises);
 
-    return files;
-  }, [] as string[]);
+  // Unique file names
+  return [...new Set(files.flatMap((f) => f))];
 }
 
-function findMatchingTests(pattern: string, cwd: string, relativePathToMain: string): string[] {
+const normalizePath = (path: string): string => path.replace(/\\/g, '/');
+
+async function findMatchingTests(
+  pattern: string,
+  workspaceRoot: string,
+  projectSourceRoot: string,
+): Promise<string[] | string> {
```

Comment:
Can we have this always return `string[]` and just collapse a single result to `[ result ]`?